### PR TITLE
Initializer calling fixed for multiworkspace

### LINKED ISSUE #1192: Optimizer job is failing
Its Observed that the Optimizer job is failing.

### PR DESCRIPTION
**Pull request by sriram251-code**
closes #1192 

Previously orgID,ApiURL and Cloud all are derived automatically.
Now we are getting the API URL from config table and OrgID from pipeline report.
Now while calling the Initializer function from Optimizer we are passing the OrgID and Api URL .

### CODE REVIEW

**Review by aman-db — APPROVED** [review]

File: src/main/scala/com/databricks/labs/overwatch/Optimizer.scala [hunk 1]
```diff
@@ -31,22 +31,68 @@ object Optimizer extends SparkSessionWrapper{
...
+
   /**
    * Derive the workspace from the config supplied in the last run
    * @param overwatchETLDB name of the Overwtch ETL database
    * @return workspace object
    */
   private def getLatestWorkspace(overwatchETLDB: String,orgId: String): Workspace = {
-    val params = getLatestSuccessState(overwatchETLDB,orgId)
+   val latestSuccessState = getLatestSuccessState(overwatchETLDB,orgId).limit(1)
+   val orgID = latestSuccessState.select("organization_id").collect().map(x => x(0).toString)
```
Comment (neilbest-db) [inline src/main/scala/com/databricks/labs/overwatch/Optimizer.scala:75]: . . . then here you could simply say:
```scala
 val ( orgID: Option[String], params: OverwatchParams) = latestSuccessState
```
or even combine the two statements if you don't need the `latestSuccessState` tuple anywhere else.

I did this as a test to make sure I was giving you good advice:

image

File: src/main/scala/com/databricks/labs/overwatch/Optimizer.scala [hunk 2]
```diff
@@ -31,22 +31,68 @@ object Optimizer extends SparkSessionWrapper{
...
+    val params = latestSuccessState
       .selectExpr("inputConfig.*")
       .as[OverwatchParams]
       .first
 
     val args = JsonUtils.objToJson(params).compactString
```
Comment (neilbest-db) [inline src/main/scala/com/databricks/labs/overwatch/Optimizer.scala:81]: these two calls to `JsonUtils.objToJson` seem redundant.  consider instead:

```scala
val args = JsonUtils.objToJson(params)
logger.log(Level.INFO, s"ARGS: Identified config string is: \n${args.prettyString}")
try {
  val apiURL = getApiURL( orgID(0), params.dataTarget)
```
. . .
```scala
  Initializer( args.compactString, apiURL, orgID)
```
. . . 

Then you already have the `Option` types that you need, right?

**Review by neilbest-db — COMMENTED** [review]
I made some suggestions to make your code cleaner.  Let Scala do the work!

File: src/main/scala/com/databricks/labs/overwatch/Optimizer.scala [hunk 3]
```diff
@@ -31,22 +31,68 @@ object Optimizer extends SparkSessionWrapper{
...
+
+
   /**
    * Derive the workspace from the config supplied in the last run
    * @param overwatchETLDB name of the Overwtch ETL database
    * @return workspace object
    */
   private def getLatestWorkspace(overwatchETLDB: String,orgId: String): Workspace = {
-    val params = getLatestSuccessState(overwatchETLDB,orgId)
+   val latestSuccessState = getLatestSuccessState(overwatchETLDB,orgId).limit(1)
```
[the remark this answers is not in the document]
Comment (sriram251-code) [reply src/main/scala/com/databricks/labs/overwatch/Optimizer.scala]: good point implemented

File: src/main/scala/com/databricks/labs/overwatch/Optimizer.scala (hunk 1 above)
Comment (sriram251-code) [reply src/main/scala/com/databricks/labs/overwatch/Optimizer.scala]: implemented

File: src/main/scala/com/databricks/labs/overwatch/Optimizer.scala (hunk 2 above)
Comment (sriram251-code) [reply src/main/scala/com/databricks/labs/overwatch/Optimizer.scala]: implemented

**Review by gueniai — APPROVED** [review]

**Pull request merged**